// @nuxtjs/seo/dist/const.mjs
const NuxtSEO = {
  slug: "nuxt-seo",
  label: "Nuxt SEO",
  icon: "i-carbon-3rd-party-connected",
  description: "The all-in-one module that brings it all together.",
  repo: "harlan-zw/nuxt-seo",
  npm: "@nuxtjs/seo"
};
const SiteConfigModule = {
  slug: "site-config",
  label: "Site Config",
  icon: "i-carbon-settings-check",
  description: "Powerful build and runtime shared site configuration for Nuxt modules.",
  repo: "harlan-zw/nuxt-site-config",
  npm: "nuxt-site-config"
};
const RobotsModule = {
  slug: "robots",
  label: "Robots",
  icon: "i-carbon-bot",
  description: "Tame the robots crawling and indexing your site with ease.",
  repo: "nuxt-modules/robots",
  npm: "@nuxtjs/robots"
};
const SitemapModule = {
  slug: "sitemap",
  label: "Sitemap",
  icon: "i-carbon-load-balancer-application",
  description: "Powerfully flexible XML Sitemaps that integrate seamlessly.",
  repo: "nuxt-modules/sitemap",
  npm: "@nuxtjs/sitemap"
};
const OgImageModule = {
  slug: "og-image",
  label: "OG Image",
  icon: "i-carbon-image-search",
  description: "Generate OG Images with Vue templates in Nuxt.",
  repo: "nuxt-modules/og-image",
  npm: "nuxt-og-image"
};
const LinkCheckerModule = {
  slug: "link-checker",
  label: "Link Checker",
  icon: "i-carbon-cloud-satellite-link",
  description: "Find and magically fix links that may be negatively effecting your SEO.",
  repo: "harlan-zw/nuxt-link-checker",
  npm: "nuxt-link-checker"
};
const SeoUtilsModule = {
  slug: "seo-utils",
  label: "SEO Utils",
  icon: "i-carbon-tools",
  description: "SEO utilities to improve your Nuxt sites discoverability and shareability.",
  repo: "harlan-zw/nuxt-seo-utils",
  npm: "nuxt-seo-utils"
};
const SchemaOrgModule = {
  slug: "schema-org",
  label: "Schema.org",
  icon: "i-carbon-chart-relationship",
  description: "The quickest and easiest way to build Schema.org graphs.",
  repo: "harlan-zw/nuxt-schema-org",
  npm: "nuxt-schema-org"
};
const modules = [
  NuxtSEO,
  RobotsModule,
  SitemapModule,
  OgImageModule,
  SchemaOrgModule,
  LinkCheckerModule,
  SeoUtilsModule,
  SiteConfigModule
];
const explicitModules = [
  RobotsModule,
  SitemapModule,
  OgImageModule,
  SchemaOrgModule,
  LinkCheckerModule,
  SeoUtilsModule
];

export { LinkCheckerModule, NuxtSEO, OgImageModule, RobotsModule, SchemaOrgModule, SeoUtilsModule, SiteConfigModule, SitemapModule, explicitModules, modules };
